### PR TITLE
Feature/232 update project by

### PR DESCRIPTION
**Pull request by paulo-teran-jalafund**
[AB#232](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/232)

# Overview

UpdateProjectById Added
Added tests for the new feature in API and Core

### CODE REVIEW

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        public PatchProjectByIdTest()
+        {
+            this.mockProjectsService = new Mock<IProjectsService>();
+            this.mockHttpContext = new DefaultHttpContext();
+            this.updateProject = new PatchProjectById(this.mockProjectsService.Object);
+        }
+
+        [Fact]
+        public async void UpdateProject_Returns_CreatedResult()
+        {
```
Comment (lilian-delgado-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs]: Update the test name to UpdateProjectById_IdIsValid_OkObjectResult, I think it describe better your test

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs [hunk 2]
```diff
@@ -132,5 +132,41 @@ public void DeleteProjectById_IdIsValid_ProjectRemoved()
             // Assert
             Assert.NotEqual(countBeforeDelete, countAfterDelete);
         }
+
+        [Fact]
+        public void UpdateProject_Returns_SingleUpdatedProject()
+        {
```
Comment (lilian-delgado-jalafund) [inline tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs]: Here too  UpdateProjectById_ProjectIsValid_ProjectUpdated()

**Review by lilian-delgado-jalafund — COMMENTED** [review]
left some comments about tests names, please review them

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs (hunk 1 above)
Comment (paulo-teran-jalafund) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs]: Updated it, please check it out

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs (hunk 2 above)
Comment (paulo-teran-jalafund) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs]: Also updated it

**Review by lilian-delgado-jalafund — DISMISSED** [review]
Lgtm!

**Review by alejandro-castellon — DISMISSED** [review]
LGTM

**Review by freddy-perez-jalafund — DISMISSED** [review]
Approved

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchProjectById.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        [OpenApiOperation(operationId: "UpdateProject", tags: new[] { "Projects" })]
+        [OpenApiRequestBody("application/json", typeof(JsonPatchDocument<Project>), Description = "JSON request body")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Project), Description = "Successful response")]
+        [OpenApiResponseWithoutBody(statusCode: HttpStatusCode.NotFound, Description = "Resource not found")]
+        public async Task<IActionResult> UpdateProject(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "patch", Route = "v1/projects/{id:guid}")] HttpRequest req, Guid id)
+        {
+            try
+            {
+                var projectUpd = this.updateProjectService.GetProject(id);
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchProjectById.cs]: please use a more readable name, don't use abbreviations
https://williamdurand.fr/2013/06/03/object-calisthenics/#6-dont-abbreviate
```suggestion
                var project = this.updateProjectService.GetProject(id);
```

File: src/dal/Jalasoft.TeamUp.Projects.DAL.Interfaces/IRepository.cs [hunk 4]
```diff
@@ -13,5 +14,7 @@ public interface IRepository<T>
         public T Add(T project);
 
         public void Remove(Guid id);
+
+        public T UpdateById(Project project);
```
Comment (cristian-chavez-jalafund) [inline src/dal/Jalasoft.TeamUp.Projects.DAL.Interfaces/IRepository.cs]: shouldn't this need to be generic?

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                        {
+                            Name = "Paola Quintanilla",
+                            IdResume = new Guid("536316e6-f8f6-41ea-b1ce-455b92be9303")
+                        }
+        },
+                State = true,
+                TextInvitation = "You are invited to be part of TeamUp",
+                CreationDate = DateTime.Today.AddDays(-10),
+            };
+            static Stream SetStream(string body)
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs]: please move this method to a helper or a util class to be shared across different tests

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        ]");
+            this.mockProjectsService.Setup(service => service.GetProject(new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"))).Returns(stubProject);
+            this.mockProjectsService.Setup(service => service.UpdateProject(null)).Returns(stubProject);
+            var response = await this.updateProject.UpdateProject(request, new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"));
+            var okObjectResult = Assert.IsType<OkObjectResult>(response);
+
+            Assert.Equal(200, okObjectResult.StatusCode);
+        }
+    }
+}
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs:109]: add other edge case scenarios like not found, bad request

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        private readonly PatchProjectById updateProject;
+
+        public PatchProjectByIdTest()
+        {
+            this.mockProjectsService = new Mock<IProjectsService>();
+            this.mockHttpContext = new DefaultHttpContext();
+            this.updateProject = new PatchProjectById(this.mockProjectsService.Object);
+        }
+
+        [Fact]
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs]: please use a theory to send different use successful use cases, to the same for other scenarios

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs [hunk 8]
```diff
@@ -132,5 +132,41 @@ public void DeleteProjectById_IdIsValid_ProjectRemoved()
             // Assert
             Assert.NotEqual(countBeforeDelete, countAfterDelete);
         }
+
+        [Fact]
+        public void UpdateProjectById_ProjectIsValid_ProjectUpdated()
+        {
+            // Arrange
+            var stubProject = new Project
```
Comment (cristian-chavez-jalafund) [inline tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs]: could this stub be abstracted to a method? so the unit test will be cleaner, please coul you try that?

Comment (cristian-chavez-jalafund) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/ProjectsServiceTests.cs]: add not successfull scenarios as well

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        public async void UpdateProjectById_IdIsValid_OkObjectResult()
+        {
+            var stubProject = new Project
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchProjectByIdTest.cs]: could this stub be abstracted to a method? so the unit test will be cleaner, please could you try that?

**Review by alejandro-castellon — DISMISSED** [review]

**Review by alejandro-castellon — APPROVED** [review]

**Review by lilian-delgado-jalafund — APPROVED** [review]

**Review by freddy-perez-jalafund — APPROVED** [review]
LGTM

**Review by cristian-chavez-jalafund — APPROVED** [review]
